### PR TITLE
(653) Reason for referral

### PR DESCRIPTION
**Pull request by jsrobertson**
## Context

https://trello.com/c/LSccMbIN/653-add-reason-for-the-referral-form-step-ui

## Changes in this PR
- Added page template for referral reason form step
- Added `/reason-for-referral` path
- Added `reason` method to `ReferralsController`
- Update `update` method to handle `reason`
- Add url to link to new page from task list

## Screenshots of UI changes
localhost_3000_referrals_778ad55f-aa95-48b7-9f4b-96a58c068c7e_reason

## Release checklist

[Release process documentation](../doc/how-to/perform-a-release.md)

As part of our continuous deployment strategy we must ensure that this work is
ready to be released once merged.

### Pre-merge

- [ ] There are changes required to the Accredited Programmes API for this change to work...
  - [ ] ... and they have been released to production already

### Post-merge

- [x] This adds, extends or meaningfully modifies a feature...
  - [ ] ... and I have written or updated an end-to-end test for the happy path in the [Accredited Programmes E2E repo](https://github.com/ministryofjustice/hmpps-accredited-programmes-e2e)
- [ ] This makes new expectations of the API...
  - [ ] ... and I have notified the API developer(s) of changes to the contract tests (Pact), or the API is already compliant
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-preprod-environment) release to preprod
- [ ] [Manually approve](../doc/how-to/perform-a-release.md#releasing-to-the-production-environment) release to prod

### CODE REVIEW

File: server/controllers/refer/referralsController.ts
Comment (jsrobertson) [inline server/controllers/refer/referralsController.ts:146]: Spotted some of the other methods incorrectly(?) refer to `req.params.prisonNumber` which I can update in another commit/PR.

File: server/controllers/refer/referralsController.ts
Comment (jsrobertson) [inline server/controllers/refer/referralsController.ts:267]: Might make sense to `trim` the reason to prevent any unwanted whitespace?

File: server/controllers/refer/referralsController.ts
Comment (yndajas) [reply server/controllers/refer/referralsController.ts:146]: Good catch. I think these have been missed because we aren't testing the message - only the status code. We tried before, but the assertion would pass regardless of what string you pass in

File: server/controllers/refer/referralsController.ts
Comment (yndajas) [reply server/controllers/refer/referralsController.ts:267]: Yeah, and hopefully to stop up accepting a string like `\n\n\n` becing a truthy `reason` (haven't checked what logic you have on the task list to say whether this is completed yet)

File: server/views/referrals/reason.njk
Comment (yndajas) [inline server/views/referrals/reason.njk:49]: This isn't in the design, but I think we do need a visible label of some kind, so it makes sense to include it here. I wonder if it would be better to follow the name in the heading here to make it read a bit softer: "Reason for referral and any additional information"

File: server/controllers/refer/referralsController.test.ts
Comment (yndajas) [inline server/controllers/refer/referralsController.test.ts:427]: I missed this when reviewing other pages, but we should have a ``describe('when the person service returns `null`'`` block here

File: server/paths/refer.ts
Comment (yndajas) [inline server/paths/refer.ts:20]: I think this should go after `confirmOasysPath` - currently they're in roughly chronological order

File: server/routes/refer.ts
Comment (yndajas) [inline server/routes/refer.ts:25]: Likewise here (this should go after `confirmOasys`)

File: server/paths/refer.ts
Comment (yndajas) [inline server/paths/refer.ts:20]: Interested to hear your thoughts on this, but I used `oasys-confirmed` for the `confirmOasysPath` to have it correspond with the property name (feels more RESTful to me), so we're looking at a referral `/referrals/:referralId` and from that its `/oasysConfirmed` property. Here I wonder if `/reason` would offer greater parity between entities and routes (and actions)?

File: server/testutils/factories/referral.ts
Comment (yndajas) [inline server/testutils/factories/referral.ts:11]: I saw the conversation about `status` on another PR, but I wonder whether the factory default should be a freshly created referral, pushing us to be intentional when using factories to create more complete referrals, and avoid randomness with values that have functional implications

`oasysConfirmed` defaults to `false` here, not a random boolean, for example

Unsure what the default value is based on the migration here (I guess `''` or `undefined` once it hits our service): https://github.com/ministryofjustice/hmpps-accredited-programmes-api/blob/main/src/main/resources/db/migration/V11__extend_referral_table.sql

File: server/@types/models/Referral.ts
Comment (yndajas) [inline server/@types/models/Referral.ts:9]: Technically the Nunjucks doesn't need to typecheck, but I wonder if this should go before it's used in the template (currently the first commit)

Comment (yndajas) [inline server/@types/models/Referral.ts:9]: Can we update `ReferralUtils.isReadyForSubmission` and relevant tests with this change?

File: server/controllers/refer/referralsController.ts
Comment (yndajas) [reply server/controllers/refer/referralsController.ts:267]: Yeah, and hopefully to prevent us reading a string like `\n\n\n` as a truthy `reason` (haven't checked what logic you have on the task list to say whether this is completed yet)

File: server/controllers/refer/referralsController.test.ts
Comment (yndajas) [inline server/controllers/refer/referralsController.test.ts:312]: Does the line at the end of this block not need updating too? Seems to be passing 🤔 Could be a quirk of `undefined` - not defining it === `undefined`

```ts
expect(referralService.updateReferral).toHaveBeenCalledWith(token, referral.id, { oasysConfirmed: true })
```

File: server/utils/referralUtils.test.ts
Comment (yndajas) [inline server/utils/referralUtils.test.ts:63]: Per earlier comment, I think this should be the default for the factory, so if that changes, this change can probably be reverted

File: server/utils/referralUtils.test.ts
Comment (yndajas) [inline server/utils/referralUtils.test.ts:95]: Everything is `referral` so I think this can just be `reason-tag`? There shouldn't be a conflict on the task list page

File: integration_tests/pages/refer/reason.ts
Comment (yndajas) [inline integration_tests/pages/refer/reason.ts:37]: Is it worth it/possible to add use a `data-testid` for this? I know we don't on a similar method elsewhere, but it might be good to phase them in with new PRs

File: server/controllers/refer/referralsController.ts
Comment (yndajas) [reply server/controllers/refer/referralsController.ts:146]: Good catch. I think these have been missed because we aren't testing the message - only the status code. We tried before, but the assertion would pass regardless of what string you pass in

**Review by yndajas — COMMENTED** [review]
Looking good. Just a few suggestions

File: server/paths/refer.ts
Comment (jsrobertson) [reply server/paths/refer.ts:20]: I was in 2 minds, I did go with `/reason` to begin with and then started to think ahead to if there was ever another "reason" for anything during the referral process but at the moment there isn't so I'm happy to go with just `/reason`

File: server/controllers/refer/referralsController.test.ts
Comment (jsrobertson) [reply server/controllers/refer/referralsController.test.ts:312]: If it's `undefined`, it doesn't technically need to be specified, however if we think it will avoid confusion then happy to add `reason: undefined` and continue to do so in future.

File: server/controllers/refer/referralsController.ts
Comment (jsrobertson) [reply server/controllers/refer/referralsController.ts:146]: I think this might have something to do with specifying an object as the second parameter with `userMessage`.  What's the reason for doing it this way? Looking at [https-errors](https://github.com/jshttp/http-errors) readme, it looks like the error should just be passed as the second parameter.

Comment (Gweaton) [reply server/controllers/refer/referralsController.ts:146]: You're right, I think we've got slightly confused on how we should be handling errors in the application. I think we picked up this use of an object with a `userMessage` field in it from a couple of other projects, but they're using that to display the validation error to the user. We're not doing that anywhere.

I think we should refactor either:
- The code to use this userMessage variable
- The way we handle errors to pass a string (easier to test).

I'll add a card to tidy this up.

File: server/testutils/factories/referral.ts
Comment (Gweaton) [reply server/testutils/factories/referral.ts:11]: As I said in that conversation, I don't think this is all that important, as we should be explicitly setting the values we care about in the tests where they're relevant.

I know you updated this in your other PR @yndajas and I approved that but I think that if we want to represent a new, empty referral, we should use a [trait](https://github.com/thoughtbot/fishery#adding-reusable-builders-traits-to-factories) to represent this, rather than assuming the default factory-built object will represent this early stage referral.

File: server/controllers/refer/referralsController.test.ts
Comment (Gweaton) [reply server/controllers/refer/referralsController.test.ts:312]: Do we _need_ to specify that `reason` is undefined here on line 312? Is it relevant to this section of the task list?

File: server/utils/referralUtils.test.ts
Comment (Gweaton) [reply server/utils/referralUtils.test.ts:63]: Let's use a trait to represent a `referral_started` Referral to be more explicit here. I think we can do that as a separate tidyup PR rather than blocking this one.

**Review by Gweaton — APPROVED** [review]
Ynda's done a good review here, I've left a couple of small comments but I don't think there's anything really blocking this PR from my perspective. They're more things we can tidy up in a future PR.

File: server/controllers/refer/referralsController.ts
Comment (jsrobertson) [reply server/controllers/refer/referralsController.ts]: For the sake of this PR I've added a test to check for the error with a string.  Can refactor the others in the future task

File: server/controllers/refer/referralsController.test.ts
Comment (jsrobertson) [reply server/controllers/refer/referralsController.test.ts:312]: Thought about it, probably nicer to remove the undefined and just update the test to check that `referralService.updateReferral` is called with the existing reason value.

Comment (yndajas) [reply server/controllers/refer/referralsController.test.ts]: If we're [keeping a pre-filled](https://github.com/ministryofjustice/hmpps-accredited-programmes-ui/pull/193#discussion_r1324149829) `reason` as the factory default (I'm not sure what I think anymore 😅) then I think checking that it's called with the existing value makes sense here

File: integration_tests/e2e/refer.cy.ts
Comment (yndajas) [inline integration_tests/e2e/refer.cy.ts:305]: One thing I missed last night: we'll need a corresponding test in `e2eReferDisabled`

File: server/views/referrals/reason.njk
Comment (yndajas) [reply server/views/referrals/reason.njk:49]: Should the page heading be updated to "supporting" if we're going for this?

File: server/testutils/factories/referral.ts
Comment (yndajas) [reply server/testutils/factories/referral.ts:11]: Shall we revert the `status` change from my PR and make `oasysConfirmed` a random boolean in that case?

Comment (Gweaton) [reply server/testutils/factories/referral.ts:11]: I'll add a card to add a trait, we can tackle that as part of that work

File: server/controllers/refer/referralsController.test.ts
Comment (yndajas) [inline server/controllers/refer/referralsController.test.ts:342]: Shall we do the same as with `oasysConfirmed` here - not set `oasysConfirmed` when building the referral, but check that the current value is sent to `updateReferral`?

File: server/controllers/refer/referralsController.test.ts
Comment (yndajas) [reply server/controllers/refer/referralsController.test.ts]: Related: https://github.com/ministryofjustice/hmpps-accredited-programmes-ui/pull/193/files#r1324300112

File: server/views/referrals/reason.njk
Comment (Gweaton) [reply server/views/referrals/reason.njk:49]: @yndajas we've had a quick chat about this and have made the label a page title and removed this extra one. I was actually looking at outdated content so it should actually be "additional information"

Comment (yndajas) [reply server/views/referrals/reason.njk:49]: Figma! 😅

File: server/controllers/refer/referralsController.test.ts
Comment (jsrobertson) [reply server/controllers/refer/referralsController.test.ts:342]: Think this has highlighted an issue with setting `request` values in each of the tests without resetting.  Will investigate.

**Review by Gweaton — APPROVED** [review]
Still to do, in future PRs:

- Make this field required
- Ensure this section has been filled in before being able to Check your answers
- Play this answer back on the CYA page

**Pull request merged**
